README.md
# fhir_at_rest

## More FHIR functionality in Flutter

This project builds upon work done in the basic FHIR package: https://github.com/Dokotela/fhir. It is designed to allow easier RESTful requests to FHIR APIs. Most of the basic requests are described on the [RESTful API](https://www.hl7.org/fhir/https.html) page of [HL7 FHIR](https://hl7.org/fhir/), or on the page for [Search parameters](https://www.hl7.org/fhir/search.html). Not all search parameters are implemented yet, but all FHIR interactions and operations are supported.


## It's not complicated, but it is detailed

The basic idea is simple
```
VERB [base]/[type]/[id] {?_format=[mime-type]}
```
The base is whatever website you want to query (for all examples on this page, we're going to use HAPI's public server: https://hapi.fhir.org/baseR4), type is the [Resource Type](https://hl7.org/fhir/resourcelist.html), and id is the id of the resource you're intersted in. The full FHIR spec defines three formal mime-types: 
* ```application/fhir+xml```
* ```application/fhir+json```
* ```application/fhir+turtle``` (No, no one knows what the hell turtle is).

However, for this package we only allow ```application/fhir+json```. Support for other mime-types may be added in future updates.

## Types of allowed interactions (NOTE: these are *NOT* the same as RESTful requests)
* read
* vread
* update
* patch
* delete
* create
* search
* search-all
* capabilities
* transaction (aka. batch)
* history
* history-type
* history-all
* operation

## Enums

When possible enums are used and that includes resourceTypes, to ensure data quality and alignment with the specification.

### Read Request
Let's try and read Patient, id: 12345. (Side note: most of these classs are (freezed)[https://pub.dev/packages/freezed] unions. Meaning they function the same for all versions of FHIR, just replace the R4 below with dstu2, stu3, or r5).
```
final req = ReadRequest.r4(
        base: Uri.parse('https://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        id: Id('12345'),
      );
      final makeReq = await req.request();
```
Which ends up with the following request: 
```
GET https://hapi.fhir.org/baseR4/Patient/12345?_format=application/fhir+json
```

I'll mention what happens with this request and how to handle the response below. There are two other options that can be added to most requests. ```_pretty``` and ```_summary```. ```_pretty``` requests the response to be nicely formatted for human readability, generally used for debugging and development. ```_summary``` is a predefined short form of the resource in response. 
Example:
```
  final req2 = ReadRequest.r4(
    base: Uri.parse('https://hapi.fhir.org/baseR4'),
    type: R4Types.patient,
    id: Id('12345'),
    pretty: true,
    summary: Summary.true_, //this is an enum
  );
  final makeReq2 = await req2.request();

```
Result:
```
GET https://hapi.fhir.org/baseR4/Patient/12345?_format=application/fhir+json&_pretty=true&_summary=true
```

### Vread Request

This is very similar to a read request. The difference is that we specify which version of the resource we want.
```
  var req2 = VreadRequest.r4(
    base: Uri.parse('https://hapi.fhir.org/baseR4'),
    type: R4Types.patient,
    id: Id('12345'),
    versionId: Id('6789'),
  );
  var makeReq2 = await req2.request();
```
Result:
```
GET https://hapi.fhir.org/baseR4/Patient/12345/_history/6789?_format=application/fhir+json
```

### Update
This is obviously different because we're sending data instead of asking for it. We use ```PUT```. We also pass in a resource.
```
  var req4 = UpdateRequest.r4(
    base: Uri.parse('https://hapi.fhir.org/baseR4'),
    type: R4Types.patient,
    id: Id('12345'),
  );

  //if the resource id above and the one below do not match, the request function will return a failure
  var patientToUpdate = r4.Patient(resourceType: 'Patient', id: Id('12345'));
  var makeReq4 = await req4.request(patientToUpdate);
```
Result:
```
PUT https://hapi.fhir.org/baseR4/Patient/12345?_format=application/fhir+json
```
This function also passes the resource as a Map to the MakeRequest Function. All types of requests get passed to this function in order to obtain a result, and it will be discussed later.

### Patch
Same request format as Update, but using ```PATCH```, and also requiring a resource.
```
  var req5 = PatchRequest.r4(
    base: Uri.parse('https://hapi.fhir.org/baseR4'),
    type: R4Types.patient,
    id: Id('12345'),
  );
  var patientToPatch = r4.Patient(resourceType: 'Patient', id: Id('12345'));
  var makeReq5 = await req5.request(patientToPatch);
```
Result:
```
PATCH https://hapi.fhir.org/baseR4/Patient/12345?_format=application/fhir+json
```

### Delete
This one is pretty straightforward. Just need the base, type and id, and it will perform a ```DELETE``` request.
```
  var req6 = DeleteRequest.r4(
    base: Uri.parse('https://hapi.fhir.org/baseR4'),
    type: R4Types.patient,
    id: Id('12345'),
  );
  var makeReq6 = await req6.request();
```
Result:
```
DELETE https://hapi.fhir.org/baseR4/Patient/12345?_format=application/fhir+json
```
### Create
Same request format as Update, but using ```POST```, and also requiring a resource.
```
  var req7 = CreateRequest.r4(
    base: Uri.parse('https://hapi.fhir.org/baseR4'),
    type: R4Types.patient,
  );
  var patientToCreate = r4.Patient(resourceType: 'Patient', id: Id('12345'));
  var makeReq7 = await req7.request(patientToCreate);
```
Result:
```
POST https://hapi.fhir.org/baseR4/Patient?_format=application/fhir+json
```
One important thing to note on create is that no ID is provided in the request, and the server will ignore the id that is in the resource. The server will create a new Id for this resource when it is created and that will replace ```'12345'```. (I'm actually still working on trying to be able to update the version number and id offline).

### Capabilities
This is to request the server's capabilities. It uses a ```GET``` and can pass in a ```Mode``` parameter. ```Mode``` may be ```full```, ```normative```, or ```terminology```. If none is passed, it will default to ```full```.

```
  var req9 = CapabilitiesRequest.r4(
    base: Uri.parse('https://hapi.fhir.org/baseR4'),
    mode: Mode.normative,
  );
  var makeReq9 = await req9.request();
```
Result:
```
https://hapi.fhir.org/baseR4/metadata?mode=normative&_format=application/fhir+json
```
### Batch
Batch and Transactions function very similarly. They both use ```POST```. They both ONLY accept a Bundle as a resource. Before sending, they both ensure that the resource is a bundle, that the bundle type is either Batch or Transaction (respectively) and will also check and each entry in the bundle has a request and that request has a method. If these conditions are not true, the request will be denied at the server, so it's just easier to check and not waste the bandwidth.
```
  var req10 = BatchRequest.r4(
    base: Uri.parse('https://hapi.fhir.org/baseR4'),
  );
  var newBundle = r4.Bundle(
      resourceType: 'Bundle', type: r4.BundleType.batch, id: Id('12345'));
  var makeReq10 = await req10.request(newBundle);
```
This succeeds in creating an empty bundle, with the following request:
```
POST https://hapi.fhir.org/baseR4?_format=application/fhir+json
```
However, this will fail because the entry does not have a request:
```
  newBundle = r4.Bundle(
      resourceType: 'Bundle',
      type: r4.BundleType.batch,
      id: Id('12345'),
      entry: [r4.BundleEntry()]);
  makeReq10 = await req10.request(newBundle);
```
This will also fail because the request does not have a method:
```
  newBundle = r4.Bundle(
      resourceType: 'Bundle',
      type: r4.BundleType.batch,
      id: Id('12345'),
      entry: [r4.BundleEntry(request: r4.BundleRequest())]);
  makeReq10 = await req10.request(newBundle);
```
However this one will succeed:
```
  var r4Bundle = r4.Bundle(
      resourceType: 'Bundle',
      type: r4.BundleType.batch,
      id: Id('12345'),
      entry: [
        r4.BundleEntry(
            request: r4.BundleRequest(method: r4.BundleRequestMethod.delete))
      ]);
  var r4Req = await r4Batch.request(r4Bundle);
```

### Transaction
Transactions are almost the same as batches. The form and function is almost equivalent. The difference is on the server side. The actions performed on a batch request are performed independently (so some may succeed even if others fail), but a transaction request is treated as a single atomic change, so either the entire request succeeds or it doesn't.
```
  var transaction = TransactionRequest.r4(
    base: Uri.parse('https://hapi.fhir.org/baseR4'),
  );
  var transactionBundle = r4.Bundle(
      resourceType: 'Bundle',
      type: r4.BundleType.transaction,
      id: Id('12345'),
      entry: [
        r4.BundleEntry(
            request: r4.BundleRequest(method: r4.BundleRequestMethod.delete))
      ]);
  var transactionReq = await transaction.request(transactionBundle);
```
Result:
```
https://hapi.fhir.org/baseR4?_format=application/fhir+json
```
 
### History
This interaction retrieves the history of a single resource (arguments: base, type, and id), all resources of that type (arguments: base, type), or all resources supported by the system (arguments: base). These all use a ```GET```. 
```
  var req11 = HistoryRequest.r4(
    base: Uri.parse('https://hapi.fhir.org/baseR4'),
    type: R4Types.observation, //valid to remove this
    id: Id('12345'),           //valid to remove this, if no type, id will be ignored
  );
  var makeReq11 = await req11.request();
```
Result:
```
https://hapi.fhir.org/baseR4/Observation/12345/_history&_format=application/fhir+json
```
There are a number of parameters that can also be passed to a history request. These are ```_count```, ```_since```, ```_at```, and ```_list```. ```_count``` is an integer and defines the number of search results per page. ```_since``` is an Instant and will request only versions created on or after the value given. ```_at``` requests only resources that were current at some point in time specified. ```_list``` allows a specification of specific resources.
```
  var req14 = HistoryRequest.r4(
    base: Uri.parse('https://hapi.fhir.org/baseR4'),
    type: R4Types.observation,
    id: Id('12345'),
    count: 10,
    since: Instant(DateTime.now().toUtc()),
  );
  var makeReq14 = await req14.request();
```
Result:
```
https://hapi.fhir.org/baseR4/Observation/12345/_history&_format=application/fhir+json&_count=10&_since=2020-10-02T14:28:34.714205Z
```
### HEAD - same as get requests, but I haven't implemented them. Not sure I will.

#### ToDo: Variant Search

## Search

Searching is challenging. I've tried to detail it by showing examples how you would perform all of the searches listed on the [HL7 page](https://www.hl7.org/fhir/search.html). Note - all searches, like above requests, will be of formatted as fhir+json. For each type of resource there are some common fields that can be searched on. ```_content, _id, _lastUpdated, _profile, _query, _security, _source, _tag```. 

ToDo: text/filter

Basic formatting for the search request is as follows:
```
  var request = SearchRequest.r4(
    base: //base fhir URL,

CHANGELOG.md
## [0.0.8]

* Changed transaction from ```GET``` request to ```POST``` request

## [0.0.7]

* Added ability to pass headers with requests (necessary for most authentication)

## [0.0.6]

* Complete alignment to [FHIR HTTP spec](https://www.hl7.org/fhir/http.html) for allowed FHIR interactions and http verb options.
* Add support for POST and formData on SearchRequest and OperationRequest classes with the `usePost` and `useFormData` flags.
* Update to `fhir: 0.0.11`
* Add parameter encoding for all parameters including general, interaction specific, search, and custom parameters
* Add http client override to support testing and allow direct management of the client
* Add full support for trial use conditional interaction modes

## [0.0.5]

* Big thanks to [drcdev](https://github.com/drcdev) he's done all of the work for this most recent update which has brought some great changes and added functionality
* Updated the _format parameter to be encoded
* Updated tests to run using the flutter_test suite
* Add globals module under /helpers
* Single variable to keep track of test mode, set to true at the beginning of test, skips https calls, returns query string
* Creation of FHIRURI class
* Undo change to flutter_test from test
* Add parameter encoding for all parameters including general, interaction specific, search, and custom parameters
* Add http client override to support testing and allow direct management of the client
* Add full support for trial use conditional interaction modes
* Update to `fhir: 0.0.11`

## [0.0.4]

* Fixed some if-else brackets
* ran formatter
* generated documentation

## [0.0.3]

* Still an issue with the links

## [0.0.2]

* Updated pubspec.yaml
* changed to secure links

## [0.0.1]

* Created 2020-10-15

example/example.dart
import 'package:fhir/primitive_types/primitive_types.dart';
import 'package:fhir/r4.dart';
import 'package:fhir_at_rest/enums/enums.dart';
import 'package:fhir_at_rest/requests/request_types.dart';
import 'package:fhir_at_rest/resource_types/resource_types.dart';
import 'package:fhir_at_rest/search_parameters/r4/parameters.dart';
import 'package:fhir_at_rest/search_parameters/search_parameter_types/search_parameter_types.dart';
import 'package:test/test.dart';

void main() {
  group('Read Request', () {
    test('Basic Read Request', () async {
      final req = ReadRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        id: Id('12345'),
      );
      final makeReq = await req.request();
      expect(makeReq.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/Patient/12345?_format=application/fhir+json');
    });

    test('Pretty read request', () async {
      final req = ReadRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        id: Id('12345'),
        pretty: true,
      );
      final makeReq = await req.request();
      expect(makeReq.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/Patient/12345?_format=application/fhir+json&_pretty=true');
    });

    test('Pretty read request, true summary', () async {
      final req = ReadRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        id: Id('12345'),
        pretty: true,
        summary: Summary.true_,
      );
      final makeReq = await req.request();
      expect(makeReq.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/Patient/12345?_format=application/fhir+json&_pretty=true&_summary=true');
    });

    test('Request, summary count', () async {
      final req = ReadRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        id: Id('12345'),
        summary: Summary.count,
      );
      final makeReq = await req.request();
      expect(makeReq.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/Patient/12345?_format=application/fhir+json&_summary=count');
    });
  });

  group('Vread Request', () {
    test('Basic Read Request', () async {
      final req = VreadRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        id: Id('12345'),
        versionId: Id('6789'),
      );
      final makeReq = await req.request();
      expect(makeReq.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/Patient/12345/_history/6789?_format=application/fhir+json');
    });

    test('Pretty read request', () async {
      final req = VreadRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        id: Id('12345'),
        versionId: Id('6789'),
        pretty: true,
      );
      final makeReq = await req.request();
      expect(makeReq.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/Patient/12345/_history/6789?_format=application/fhir+json&_pretty=true');
    });

    test('Pretty read request, true summary', () async {
      final req = VreadRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        id: Id('12345'),
        versionId: Id('6789'),
        pretty: true,
        summary: Summary.true_,
      );
      final makeReq = await req.request();
      expect(makeReq.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/Patient/12345/_history/6789?_format=application/fhir+json&_pretty=true&_summary=true');
    });

    test('Request, summary count', () async {
      final req = VreadRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        id: Id('12345'),
        versionId: Id('6789'),
        summary: Summary.count,
      );
      final makeReq = await req.request();
      expect(makeReq.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/Patient/12345/_history/6789?_format=application/fhir+json&_summary=count');
    });
  });

  group('Batch/Transaction Requests', () {
    test('Batch Request', () async {
      final req10 = TransactionRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
      );
      final newBundle = Bundle(
          resourceType: 'Bundle', type: BundleType.batch, id: Id('12345'));
      final makeReq10 = await req10.request(newBundle);
      expect(makeReq10.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4?_format=application/fhir+json');
    });

    test('Batch Request with error', () async {
      final req21 = TransactionRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
      );
      final newBundle = Bundle(
          resourceType: 'Bundle',
          type: BundleType.batch,
          id: Id('12345'),
          entry: [BundleEntry(request: BundleRequest())]);
      final makeReq21 = await req21.request(newBundle);
      expect(makeReq21.fold((l) => l.errorMessage(), (r) => r),
          'Each bundle entry request needs a method type specified, but some of the entries in this bundle are missing a method.');
    });

    test('Transaction Request', () async {
      final transaction = TransactionRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
      );
      final transactionBundle = Bundle(
          resourceType: 'Bundle',
          type: BundleType.transaction,
          id: Id('12345'),
          entry: [
            BundleEntry(
                request: BundleRequest(method: BundleRequestMethod.delete))
          ]);
      final transactionReq = await transaction.request(transactionBundle);
      expect(transactionReq.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4?_format=application/fhir+json');
    });
  });

  group('History Requests', () {
    test('History of Specific Observation', () async {
      final req11 = HistoryRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.observation,
        id: Id('12345'),
      );
      final makeReq11 = await req11.request();
      expect(makeReq11.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/Observation/12345/_history?_format=application/fhir+json');
    });

    test('History of all Observations', () async {
      final req12 = HistoryTypeRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.observation,
      );
      final makeReq12 = await req12.request();
      expect(makeReq12.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/Observation/_history?_format=application/fhir+json');
    });

    test('History of everything', () async {
      final req13 = HistoryAllRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
      );
      final makeReq13 = await req13.request();
      expect(makeReq13.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/_history?_format=application/fhir+json');
    });

    test('History of Observations, count of 10, after a specified date',
        () async {
      final req14 = HistoryRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.observation,
        id: Id('12345'),
        count: 10,
        since: Instant('2020-10-08T16:58:07.241117Z'),
      );
      final makeReq14 = await req14.request();
      expect(makeReq14.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/Observation/12345/_history?_format=application/fhir+json&_count=10&_since=2020-10-08T16:58:07.241117Z');
    });
  });

  group('Other Requests', () {
    test('Update Request', () async {
      final req4 = UpdateRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        id: Id('12345'),
      );
      final patientToUpdate = Patient(resourceType: 'Patient', id: Id('12345'));
      final makeReq4 = await req4.request(resource: patientToUpdate);
      expect(makeReq4.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/Patient/12345?_format=application/fhir+json');
    });

    test('Patch Request', () async {
      final req5 = PatchRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        id: Id('12345'),
      );
      final patientToPatch = Patient(resourceType: 'Patient', id: Id('12345'));
      final makeReq5 = await req5.request(resource: patientToPatch);
      expect(makeReq5.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/Patient/12345?_format=application/fhir+json');
    });

    test('Delete Request', () async {
      final req6 = DeleteRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        id: Id('12345'),
      );
      final makeReq6 = await req6.request();
      expect(makeReq6.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/Patient/12345?_format=application/fhir+json');
    });

    test('Create Request', () async {
      final req7 = CreateRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
      );
      final patientToCreate = Patient(resourceType: 'Patient', id: Id('12345'));
      final makeReq7 = await req7.request(resource: patientToCreate);
      expect(makeReq7.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/Patient?_format=application/fhir+json');
    });

    test('Capabilities Request', () async {
      final req9 = CapabilitiesRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        mode: Mode.normative,
      );
      final makeReq9 = await req9.request();
      expect(makeReq9.fold((l) => l.errorMessage(), (r) => r),
          'http://hapi.fhir.org/baseR4/metadata?mode=normative&_format=application/fhir+json');
    });
  });

  group('Search Request', () {
    test('Patient ID Search', () async {
      final request = SearchRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        parameters: PatientSearch(searchId: [Id('12345')]),
      );
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Patient?_format=application/fhir+json&_id=12345',
      );
    });

    test('Observation Time Search', () async {
      final request = SearchRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.observation,
        parameters: ObservationSearch(searchLastUpdated: [
          SearchDate(date: FhirDateTime('2010-10-01'), prefix: DatePrefix.gt)
        ]),
      );
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Observation?_format=application/fhir+json&_lastUpdated=gt2010-10-01',
      );
    });

    test('Condition search, code parameter (_tag)', () async {
      final request = SearchRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.condition,
        parameters: ConditionSearch(
          searchTag: [
            SearchToken(
                system: FhirUri('http://acme.org/codes'),
                code: Code('needs-review'))
          ],
        ),
      );
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Condition?_format=application/fhir+json&_tag=http://acme.org/codes|needs-review',
      );
    });

    test('Diagnostic Report Search, Uri parameter (_profile)', () async {
      final request = SearchRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.diagnosticreport,
        parameters: DiagnosticReportSearch(
          searchProfile: [SearchUri(uri: FhirUri('http://acme.org/codes'))],
        ),
      );
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/DiagnosticReport?_format=application/fhir+json&_profile=http://acme.org/codes',
      );
    });

    test('Patient Search if gender present', () async {
      final request = SearchRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        parameters: PatientSearch(gender: [SearchToken(missing: true)]),
      );
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Patient?_format=application/fhir+json&gender:missing=true',
      );
    });

    test('Patient Search if gender present', () async {
      final request = SearchRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        parameters: PatientSearch(gender: [SearchToken(missing: false)]),
      );
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Patient?_format=application/fhir+json&gender:missing=false',
      );
    });

    test('Patient Search for exact text match', () async {
      final request = SearchRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        parameters: PatientSearch(searchText: [
          SearchString(string: 'Stark', modifier: StringModifier.exact)
        ]),
      );
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Patient?_format=application/fhir+json&_text:exact=Stark',
      );
    });

    test('Observation Search "le"', () async {
      final request = SearchRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.observation,
        parameters: ObservationSearch(searchLastUpdated: [
          SearchDate(date: FhirDateTime('2010-10-01'), prefix: DatePrefix.le)
        ]),
      );
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Observation?_format=application/fhir+json&_lastUpdated=le2010-10-01',
      );
    });

    test('Risk Assessment Search, probability > 0.8', () async {
      final request = SearchRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.riskassessment,
        parameters: RiskAssessmentSearch(
            probability: [SearchNumber(number: 0.8, prefix: NumberPrefix.gt)]),
      );
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/RiskAssessment?_format=application/fhir+json&probability=gt0.8',
      );
    });

    test('Patient Search for birthday between two dates', () async {
      final request = SearchRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.patient,
        parameters: PatientSearch(birthdate: [
          SearchDate(date: FhirDateTime('2010-01-01'), prefix: DatePrefix.ge),
          SearchDate(date: FhirDateTime('2011-12-31'), prefix: DatePrefix.le)
        ]),
      );
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Patient?_format=application/fhir+json&birthdate=ge2010-01-01&birthdate=le2011-12-31',
      );
    });

    test('Patient Search given name with parameters', () async {
      final request = SearchRequest.r4(
          base: Uri.parse('http://hapi.fhir.org/baseR4'),
          type: R4Types.patient,
          parameters: PatientSearch(
            given: [
              SearchString(string: 'eve'),
              SearchString(string: 'eve', modifier: StringModifier.contains),
              SearchString(string: 'eve', modifier: StringModifier.exact)
            ],
          ));
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Patient?_format=application/fhir+json&given=eve&given:contains=eve&given:exact=eve',
      );
    });

    test('Patient Search with identifier', () async {
      final request = SearchRequest.r4(
          base: Uri.parse('http://hapi.fhir.org/baseR4'),
          type: R4Types.patient,
          parameters: PatientSearch(
            identifier: [
              SearchToken(
                  system: FhirUri('http://acme.org/patient'),
                  code: Code('2345'))
            ],
          ));
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Patient?_format=application/fhir+json&identifier=http://acme.org/patient|2345',
      );
    });

    test('Composition Search Section Code', () async {
      final request = SearchRequest.r4(
          base: Uri.parse('http://hapi.fhir.org/baseR4'),
          type: R4Types.composition,
          parameters: CompositionSearch(
            section: [
              SearchToken(code: Code('48765-2'), modifier: TokenModifier.not)
            ],
          ));
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Composition?_format=application/fhir+json&section:not=48765-2',
      );
    });

    test('Condition search via snomed code', () async {
      final request = SearchRequest.r4(
          base: Uri.parse('http://hapi.fhir.org/baseR4'),
          type: R4Types.condition,
          parameters: ConditionSearch(
            code: [
              SearchToken(
                  system: FhirUri(
                      'http%3A%2F%2Fsnomed.info%2Fsct%3Ffhir_vs%3Disa%2F126851005'),
                  modifier: TokenModifier.in_)
            ],
          ));
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Condition?_format=application/fhir+json&code:in=http%3A%2F%2Fsnomed.info%2Fsct%3Ffhir_vs%3Disa%2F126851005',
      );
    });

    test('Patient search by mr code', () async {
      final request = SearchRequest.r4(
          base: Uri.parse('http://hapi.fhir.org/baseR4'),
          type: R4Types.patient,
          parameters: PatientSearch(
            identifier: [
              SearchToken(
                system:
                    FhirUri('http://terminology.hl7.org/CodeSystem/v2-0203'),
                code: Code('MR'),
                value: '446053',
                modifier: TokenModifier.of_type,
              ),
            ],
          ));
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Patient?_format=application/fhir+json&identifier:of-type=http://terminology.hl7.org/CodeSystem/v2-0203|MR|446053',
      );
    });

    test('Observation search by quantity, with prefix', () async {
      final request = SearchRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.observation,
        parameters: ObservationSearch(
          valueQuantity: [
            SearchQuantity(
              number: 5.4,
              system: FhirUri('http://unitsofmeasure.org'),
              code: Code('mg'),
              prefix: QuantityPrefix.ap,
            ),
          ],
        ),
      );
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Observation?_format=application/fhir+json&value-quantity=ap5.4|http://unitsofmeasure.org|mg',
      );
    });

    test('Observation search by quantity, without prefix', () async {
      final request = SearchRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.observation,
        parameters: ObservationSearch(
          valueQuantity: [
            SearchQuantity(
              number: 5.4,
              system: FhirUri('http://unitsofmeasure.org'),
              code: Code('mg'),
            ),
          ],
        ),
      );
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Observation?_format=application/fhir+json&value-quantity=5.4|http://unitsofmeasure.org|mg',
      );
    });

    test('Observation search by quantity, testing number parsing', () async {
      final request = SearchRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.observation,
        parameters: ObservationSearch(
          valueQuantity: [
            SearchQuantity(
              number: '5.4e-3',
              system: FhirUri('http://unitsofmeasure.org'),
              code: Code('g'),
            ),
          ],
        ),
      );
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Observation?_format=application/fhir+json&value-quantity=0.0054|http://unitsofmeasure.org|g',
      );
    });

    test('Observation search by reference, using url', () async {
      final request = SearchRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.observation,
        parameters: ObservationSearch(
          subject: [
            SearchReference(
              url: FhirUri('Patient/123'),
            ),
          ],
        ),
      );
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Observation?_format=application/fhir+json&subject=Patient/123',
      );
    });

    test('Observation search by reference, using type and ID', () async {
      final request = SearchRequest.r4(
        base: Uri.parse('http://hapi.fhir.org/baseR4'),
        type: R4Types.observation,
        parameters: ObservationSearch(
          subject: [
            SearchReference(
              type: R4Types.patient,
              id: Id('123'),
            ),
          ],
        ),
      );
      final response = await request.request();
      expect(
        response.fold((l) => l.errorMessage(), (r) => r),
        'http://hapi.fhir.org/baseR4/Observation?_format=application/fhir+json&subject=Patient/123',
      );
    });
  });
}
